tests/test_pkg.py

```diff
@@ -1,35 +1,77 @@
+"""Check the content of source code and source distribution.
+
+Binary distribution (wheel) is checked in test workflow.
+"""
+
 from __future__ import annotations
 
-import os
+import subprocess
+import tarfile
 from glob import glob
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from monty.tempfile import ScratchDir
+
+if TYPE_CHECKING:
+    from pymatgen.util.typing import PathLike
+
+PROJECT_ROOT = Path(__file__).parent.parent
+
+
+class TestCheckDistribution:
+    def test_source_code(self):
+        """Directly check the source code in the working directory."""
+        src_txt_path = PROJECT_ROOT / "src/pymatgen.egg-info/SOURCES.txt"
 
-import pytest
+        _check_src_txt_is_complete(PROJECT_ROOT, src_txt_path)
 
-SRC_TXT_PATH = "src/pymatgen.egg-info/SOURCES.txt"
+    def test_source_distribution(self):
+        """Build the source distribution and verify its contents."""
 
+        with ScratchDir("."):
+            # Build the source distribution
+            subprocess.run(["python", "-m", "pip", "install", "--upgrade", "build"], check=True)
+            subprocess.run(["python", "-m", "build", "--sdist", PROJECT_ROOT, "--outdir", ".", "-C--quiet"], check=True)
 
-@pytest.mark.skipif(
-    not os.path.isfile(SRC_TXT_PATH),
-    reason=f"{SRC_TXT_PATH=} not found. Run `pip install .` to create",
-)
-def test_egg_sources_txt_is_complete():
-    """Check that all source and data files in pymatgen/ are listed in pymatgen.egg-info/SOURCES.txt."""
+            # Decompress sdist
+            sdist_file = next(Path(".").glob("*.tar.gz"))
+            sdist_dir = sdist_file.name.removesuffix(".tar.gz")
+            with tarfile.open(sdist_file, "r:gz") as tar:
+                # TODO: remove attr check after only 3.12+
+                if hasattr(tarfile, "data_filter"):
+                    tar.extractall("", filter="data")
+                else:
+                    tar.extractall("")  # noqa: S202
 
-    with open(SRC_TXT_PATH, encoding="utf-8") as file:
-        sources = file.read()
+            # Verify source distribution contents
+            src_txt_path = f"{sdist_dir}/src/pymatgen.egg-info/SOURCES.txt"
```

Comment:
I'm not sure if that makes sense at all to compare sdist with `SOURCES.txt` (the contents of sdist is defined by `SOURCES.txt`), need to double [check the definition](https://setuptools.pypa.io/en/stable/deprecated/commands.html#egg-info-create-egg-metadata-and-set-build-tags):
> This command performs two operations: it updates a project’s .egg-info metadata directory (used by the bdist_egg, develop, and test commands), and it allows you to temporarily change a project’s version string, to support “daily builds” or “snapshot” releases. It is run automatically by the sdist, bdist_egg, develop, and test commands in order to update the project’s metadata, but you can also specify it explicitly in order to temporarily change the project’s version string while executing other commands. (It also generates the .egg-info/SOURCES.txt manifest file, which is used when you are building source distributions.)